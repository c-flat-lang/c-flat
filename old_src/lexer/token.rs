pub type Span = std::ops::Range<usize>;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Token {
    pub kind: TokenKind,
    pub lexeme: String,
    pub span: Span,
}

impl Token {
    pub fn new(kind: TokenKind, lexeme: impl Into<String>, span: Span) -> Self {
        Self {
            kind,
            lexeme: lexeme.into(),
            span,
        }
    }

    pub fn is_keyword(&self, expected: Keyword) -> bool {
        matches!(&self.kind, TokenKind::Keyword(actual) if actual == &expected)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenKind {
    Percent,
    Bang,
    BangEqual,
    Char,
    Comma,
    Colon,
    Dot,
    Equal,
    EqualEqual,
    Float,
    Greater,
    GreaterEqual,
    Identifier,
    InvalidToken,
    Keyword(Keyword),
    LeftBrace,
    LeftBracket,
    LeftParen,
    Less,
    LessEqual,
    Minus,
    Number,
    Plus,
    RightBrace,
    RightBracket,
    RightParen,
    Semicolon,
    Slash,
    Star,
    String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Keyword {
    Const,
    Else,
    False,
    Fn,
    For,
    If,
    Let,
    Pub,
    Return,
    Struct,
    True,
    Type,
    Use,
    While,
}
