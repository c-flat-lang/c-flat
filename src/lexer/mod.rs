pub mod token;
pub mod tokenizer;

pub fn tokenize(source: &str) -> Vec<token::Token> {
    tokenizer::Tokenizer::new(source).collect()
}
