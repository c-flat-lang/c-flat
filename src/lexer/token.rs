pub type Span = std::ops::Range<usize>;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Token {
    pub kind: TokenKind,
    pub lexeme: String,
    pub span: Span,
}

impl Token {
    pub fn new(kind: TokenKind, lexeme: impl Into<String>, span: Span) -> Self {
        Self {
            kind,
            lexeme: lexeme.into(),
            span,
        }
    }

    pub fn is_keyword(&self, expected: Keyword) -> bool {
        matches!(&self.kind, TokenKind::Keyword(actual) if actual == &expected)
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TokenKind {
    Identifier,
    Keyword(Keyword),
    Number,
    Float,
    String,
    Char,
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    LeftBracket,
    RightBracket,
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,
    Bang,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,
    InvalidToken,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Keyword {
    Pub,
    Fn,
    Let,
    Const,
    If,
    Else,
    While,
    For,
    Return,
}
