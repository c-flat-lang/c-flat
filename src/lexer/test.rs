use crate::lexer;
use std::fmt::Write;

pub fn snapshot_lexing(input: &str) -> String {
    let tokens = lexer::tokenize(input);
    let mut tokens = std::collections::VecDeque::from(tokens);
    let mut output = String::new();
    let mut total = 0;
    for line in input.lines() {
        output += line;
        output += "\n";
        while let Some(tok) = tokens.pop_front() {
            if total + line.len() <= tok.span.start {
                tokens.push_front(tok);
                break;
            }

            output += &" ".repeat(tok.span.start.saturating_sub(total));
            output += &"^".repeat(tok.span.len());
            write!(&mut output, " {tok:?}").expect("failed to write()");
            output += "\n"
        }

        total += line.len() + 1;
    }

    output
}

macro_rules! snapshot {
    ($name:tt, $path:tt) => {
        #[test]
        fn $name() {
            let contents = include_str!($path);
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("testdata/output/");
            settings.bind(|| {
                insta::assert_snapshot!(snapshot_lexing(contents));
            });
        }
    };
}

snapshot!(hello_world, "../../examples/hello_world.cb");
snapshot!(char, "../../examples/char.cb");
