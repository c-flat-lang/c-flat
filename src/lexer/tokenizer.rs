use super::token::{Keyword, Span, Token, TokenKind};
pub struct Tokenizer<'a> {
    chars: std::iter::Peekable<std::str::Chars<'a>>,
    span: Span,
}

impl<'a> Tokenizer<'a> {
    pub fn new(src: &'a str) -> Self {
        Self {
            chars: src.chars().peekable(),
            span: 0..0,
        }
    }

    fn next_char(&mut self) -> Option<char> {
        let value = self.chars.next();
        if value.is_some() {
            self.span.end += 1;
        }
        value
    }

    fn next_char_if(&mut self, predicate: impl Fn(char) -> bool) -> Option<char> {
        match self.chars.peek() {
            Some(value) if predicate(*value) => self.next_char(),
            _ => None,
        }
    }

    fn peek_char(&mut self, expected: char) -> bool {
        matches!(self.chars.peek(), Some(actual) if actual == &expected)
    }

    fn spanned(&mut self, kind: TokenKind, lexeme: impl Into<String>) -> Token {
        let span = self.span.clone();
        self.span = self.span.end..self.span.end;
        Token::new(kind, lexeme, span)
    }

    fn parse_number(&mut self, value: char) -> Token {
        let mut lexeme = String::from(value);

        while let Some(value) =
            self.next_char_if(|value| value.is_ascii_digit() || ['.', '_'].contains(&value))
        {
            lexeme.push(value);
        }

        if lexeme.contains('.') {
            return self.spanned(TokenKind::Float, lexeme);
        }

        self.spanned(TokenKind::Number, lexeme)
    }

    fn parse_identifier(&mut self, value: char) -> Token {
        let mut lexeme = String::from(value);
        while let Some(value) =
            self.next_char_if(|value| value.is_ascii_alphanumeric() || value == '_')
        {
            lexeme.push(value);
        }

        let kind = match lexeme.as_str() {
            "const" => TokenKind::Keyword(Keyword::Const),
            "else" => TokenKind::Keyword(Keyword::Else),
            "false" => TokenKind::Keyword(Keyword::False),
            "fn" => TokenKind::Keyword(Keyword::Fn),
            "for" => TokenKind::Keyword(Keyword::For),
            "if" => TokenKind::Keyword(Keyword::If),
            "let" => TokenKind::Keyword(Keyword::Let),
            "pub" => TokenKind::Keyword(Keyword::Pub),
            "return" => TokenKind::Keyword(Keyword::Return),
            "true" => TokenKind::Keyword(Keyword::True),
            "while" => TokenKind::Keyword(Keyword::While),
            _ => TokenKind::Identifier,
        };
        self.spanned(kind, lexeme)
    }

    fn parse_string(&mut self) -> Token {
        let mut lexeme = String::from('#');
        while let Some(value) = self.next_char() {
            lexeme.push(value);
            if lexeme.ends_with("\"#") {
                break;
            }
        }
        let lexeme = lexeme[2..lexeme.len() - 2].replace("\\n", "\n");
        self.spanned(TokenKind::String, lexeme)
    }

    fn parse_char(&mut self) -> Token {
        let mut lexeme = String::from('#');
        while let Some(value) = self.next_char() {
            lexeme.push(value);
            if lexeme.ends_with("\'#") {
                break;
            }
        }
        let lexeme = lexeme[2..lexeme.len() - 2].replace("\\n", "\n");
        self.spanned(TokenKind::Char, lexeme)
    }

    fn double_char(&mut self, kind: TokenKind, lexeme: impl Into<String>) -> Option<Token> {
        self.next_char();
        Some(self.spanned(kind, lexeme.into()))
    }

    fn skip_char(&mut self) -> Option<Token> {
        self.spanned(TokenKind::InvalidToken, ' ');
        self.next()
    }
}

impl<'a> Iterator for Tokenizer<'a> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        let Some(c) = self.next_char() else {
            return None;
        };
        match c {
            '0'..='9' => Some(self.parse_number(c)),
            value if value.is_ascii_alphabetic() => Some(self.parse_identifier(value)),
            value if value.is_ascii_whitespace() => self.skip_char(),
            '#' if self.peek_char('\'') => Some(self.parse_char()),
            '#' if self.peek_char('"') => Some(self.parse_string()),
            '=' if self.peek_char('=') => Some(self.spanned(TokenKind::EqualEqual, c)),
            '>' if self.peek_char('=') => Some(self.spanned(TokenKind::GreaterEqual, c)),
            '<' if self.peek_char('=') => Some(self.spanned(TokenKind::LessEqual, c)),
            '!' if self.peek_char('=') => self.double_char(TokenKind::BangEqual, "!="),
            '(' => Some(self.spanned(TokenKind::LeftParen, c)),
            ')' => Some(self.spanned(TokenKind::RightParen, c)),
            '[' => Some(self.spanned(TokenKind::LeftBracket, c)),
            ']' => Some(self.spanned(TokenKind::RightBracket, c)),
            '{' => Some(self.spanned(TokenKind::LeftBrace, c)),
            '}' => Some(self.spanned(TokenKind::RightBrace, c)),
            ',' => Some(self.spanned(TokenKind::Comma, c)),
            '.' => Some(self.spanned(TokenKind::Dot, c)),
            '-' => Some(self.spanned(TokenKind::Minus, c)),
            '+' => Some(self.spanned(TokenKind::Plus, c)),
            ';' => Some(self.spanned(TokenKind::Semicolon, c)),
            ':' => Some(self.spanned(TokenKind::Colon, c)),
            '/' => Some(self.spanned(TokenKind::Slash, c)),
            '*' => Some(self.spanned(TokenKind::Star, c)),
            '!' => Some(self.spanned(TokenKind::Bang, c)),
            '=' => Some(self.spanned(TokenKind::Equal, c)),
            '>' => Some(self.spanned(TokenKind::Greater, c)),
            '<' => Some(self.spanned(TokenKind::Less, c)),
            c => Some(self.spanned(TokenKind::InvalidToken, c)),
        }
    }
}
