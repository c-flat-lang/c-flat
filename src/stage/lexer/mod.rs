pub mod token;
mod tokenizer;
use token::Token;

use super::Stage;

#[derive(Debug, Default)]
pub struct Lexer {}

impl Stage<&str, Vec<Token>> for Lexer {
    fn run(&mut self, input: &str) -> Vec<Token> {
        let tokenizer = tokenizer::Tokenizer::new(input);
        tokenizer.collect()
    }
}
