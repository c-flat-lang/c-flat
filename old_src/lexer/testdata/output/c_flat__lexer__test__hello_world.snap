---
source: src/lexer/test.rs
expression: snapshot_lexing(contents)
---
pub fn main() void {
^^^ Token { kind: Keyword(Pub), lexeme: "pub", span: 0..3 }
    ^^ Token { kind: Keyword(Fn), lexeme: "fn", span: 4..6 }
       ^^^^ Token { kind: Identifier, lexeme: "main", span: 7..11 }
           ^ Token { kind: LeftParen, lexeme: "(", span: 11..12 }
            ^ Token { kind: RightParen, lexeme: ")", span: 12..13 }
              ^^^^ Token { kind: Identifier, lexeme: "void", span: 14..18 }
                   ^ Token { kind: LeftBrace, lexeme: "{", span: 19..20 }
    println(#"Hello, World!\n"#);
    ^^^^^^^ Token { kind: Identifier, lexeme: "println", span: 25..32 }
           ^ Token { kind: LeftParen, lexeme: "(", span: 32..33 }
            ^^^^^^^^^^^^^^^^^^^ Token { kind: String, lexeme: "Hello, World!\n", span: 33..52 }
                               ^ Token { kind: RightParen, lexeme: ")", span: 52..53 }
                                ^ Token { kind: Semicolon, lexeme: ";", span: 53..54 }
}
^ Token { kind: RightBrace, lexeme: "}", span: 55..56 }
